src/mpid/ch4/netmod/ofi/ofi_types.h

```diff
@@ -68,42 +68,71 @@
 #endif
 
 /* RMA Key Space division
- *    |                  |                  |                    |
- *    ...     Context ID |   Huge RMA       |  Window Instance   |
- *    |                  |                  |                    |
+ *    |                  |                   |                        |
+ *    ...     Context ID | key type flag bit |  Key/window instance   |
+ *    |                  |                   |                        |
  */
 /* 64-bit key space                         */
-/* 2M  window instances per comm           */
-/* 2M  outstanding huge RMAS per comm      */
-/* 4M  communicators                       */
-#define MPIDI_OFI_MAX_WINDOWS_BITS_64  (21)
-#define MPIDI_OFI_MAX_HUGE_RMA_BITS_64 (21)
-#define MPIDI_OFI_MAX_HUGE_RMAS_64     (1<<(MPIDI_OFI_MAX_HUGE_RMA_BITS_64))
-#define MPIDI_OFI_MAX_WINDOWS_64       (1<<(MPIDI_OFI_MAX_WINDOWS_BITS_64))
-#define MPIDI_OFI_HUGE_RMA_SHIFT_64    (MPIDI_OFI_MAX_WINDOWS_BITS_64)
-#define MPIDI_OFI_CONTEXT_SHIFT_64     (MPIDI_OFI_MAX_WINDOWS_BITS_64+MPIDI_OFI_MAX_HUGE_RMA_BITS_64)
+/* 2147483648 window instances per comm           */
+/* 2147483648 outstanding huge RMAS per comm      */
+/* 4294967296 communicators                       */
+#define MPIDI_OFI_MAX_KEY_TYPE_BITS_64 (1)
+#define MPIDI_OFI_MAX_CONTEXT_BITS_64  (32)
+#define MPIDI_OFI_MAX_KEY_BITS_64      (64-(MPIDI_OFI_MAX_KEY_TYPE_BITS_64+MPIDI_OFI_MAX_CONTEXT_BITS_64))
+#define MPIDI_OFI_MAX_HUGE_RMAS_64     (1<<(MPIDI_OFI_MAX_KEY_BITS_64))
+#define MPIDI_OFI_MAX_WINDOWS_64       MPIDI_OFI_MAX_HUGE_RMAS_64
+
+#define MPIDI_OFI_CONTEXT_SHIFT_64     (MPIDI_OFI_MAX_KEY_BITS_64+MPIDI_OFI_MAX_KEY_TYPE_BITS_64)
+#define MPIDI_OFI_MAX_KEY_TYPE_SHIFT_64 (MPIDI_OFI_MAX_KEY_BITS_64)
+#define MPIDI_OFI_MAX_KEY_SHIFT_64     (0)
```

Comment:
Assuming the above layout that the key always sits in the least significant bits, we can assume key shift is always zero and thus remove this macro? (also for 32-bit and 16-bit cases.)